CAD-2147: Support additional scripts in structured metadata.

nc6 opened the pull request:
Additional comments per commit.

nc6 committed:
Generalise Metadata for different eras.

Introduce a 'ValidateMetadata' class which operates across eras.
Currently tests are still failing, since it may be easier to operate
this atop of the generalised test generation framework.

nc6 committed:
Drop unneeded shelley-ma dependency in tests.

nc6 committed:
Implement Metadata for Shelley MA.

This covers CAD-2147, and adds support for adding additional scripts to
the metadata in a structured way.

uroboros reviewed (approved):
looks good! - also looks like a painful rebase in my near future :)

JaredCorduan commented on shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Metadata.hs:
```diff
@@ -0,0 +1,143 @@
+-- | Raw, un-memoised metadata type
+data MetadataRaw era = MetadataRaw
+  { mdScriptPreimages :: !(StrictSeq (Core.Script era)),
+    -- | Unstructured metadata "blob"
```
so sorry to bike-shed a word in a comment, and a variable name.  I myself had been using the word "unstructured" for this component, but @dcoutts let me know that actually we've been elsewhere been calling this "structured metadata" and this will confuse folks. `mdScriptPreimages` is great, though. :)

JaredCorduan commented on shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Metadata.hs:
```diff
@@ -0,0 +1,143 @@
+      ( Ann (RecD MetadataRaw)
+          <*! D (sequence <$> decodeStrictSeq fromCBOR)
+          <*! Ann From
+      )
```
I thought that we wanted the deserialization to be backwards compatible.

nc6 replied:
Yes, I am an idiot and totally forgot about this. Will fix.

JaredCorduan replied:
you are anything but that! and thanks again for taking this one on!

JaredCorduan replied:
@redxaxder and I propose this: https://github.com/input-output-hk/cardano-ledger-specs/pull/1994/files#diff-03ecce724e5f30f7e4d885d88b81708a4bc1bc0451d2aa8b82137f7791f5c07bR214-R219

JaredCorduan reviewed (commented):
@nc6 this is really awesome, thank you. I made two comments, one about names :scream: and one about backwards compatibility.

nc6 commented:
I should point out that the failing CI here is expected; I'm waiting for Ryan's PR on generalising the tests to fix things.

nc6 committed:
Add tests for Metadata serialisation.

- Round-tripping of MA-era Metadata
- Check that Shelley-era metadata deserialises as MA-era.

nc6 commented:
@redxaxder Please note that in the last commit I changed the CDDL format to match the encoder, because it seemed completely non-obvious to me how to use `Coders` to represent the suggested format. But perhaps you or @TimSheard know how to do this easily, in which case I'd welcome a commit changing the representation (or just advice as to how to do this). This can either be part of this PR or a follow-up

redxaxder commented on shelley-ma/shelley-ma-test/cddl-files/shelley-ma.cddl:
```diff
@@ -211,10 +211,10 @@ transaction_metadatum_label = uint
+     , 1: [ * native_script ]
      ; other types of metadata...
-    }]
+    ]
```
If you're changing it to a dense encoding then the labels `0:` and `1:` are just informative rather than part of the encoding. IIRC some code generators emit these as names. So these could be more descriptive.

redxaxder commented:
> @redxaxder Please note that in the last commit I changed the CDDL format to match the encoder, because it seemed completely non-obvious to me how to use `Coders` to represent the suggested format. But perhaps you or @TimSheard know how to do this easily, in which case I'd welcome a commit changing the representation (or just advice as to how to do this). This can either be part of this PR or a follow-up

Your chosen encoding here is actually smaller than the one before. (In the case where there are two kinds of metadata, which is where we are). I think it might we worth talking about longer term plans for this.
